lib/account/account-manager.js

```diff
@@ -83,6 +145,147 @@ module.exports = function (accountService, config, hotp, otDate, promise, random
 
 
     /**
+     * Creates a password hash using the same method
+     * an account holder used to create a password to login.
+     *
+     * @param {string} password
+     * @param {string} salt
+     * @return {Promise.<string>}
+     */
+    function hashPasswordAsync(accountFile, salt) {
+        return secureHash.hashAsync(accountFile.password + salt, accountConfig.passwordHash.secondary, true).then((hashedPassword) => {
+            return secureHash.simpleHash(hashedPassword, accountConfig.loginHash);
+        });
+    }
+
+
+    /**
+     * Creates the object for key expiration option. We need to be
+     * specific that we are expiring a file after a given time.
+     *
+     * @param {accountManager~lifeTime} lifeTime
+     * @return {s3~putOptions}
+     * @throws {Error} if we don't pass in something or it's not an object
+     */
+    function keyExpiration(lifeTime) {
+        if (!lifeTime) {
+            throw new Error("Lifetime for key was not found");
+        }
+
+        return otDate.now().plus(lifeTime);
+    }
+
+
+    /**
+     * Finishes logging in a user.
+     *
+     * @param {string} hashedAccountId
+     * @param {string} loginId
+     * @param {accountManager~loginCompleteInfo} loginInfo
+     * @return {Promise.<*>}
+     */
+    function loginCompleteAsync(hashedAccountId, loginId, loginInfo) {
+        return accountService.getLoginFileAsync(hashedAccountId, loginId).then((loginFile) => {
+            return accountService.getAccountFileAsync(loginFile.accountId).then((accountFile) => {
+                return loginVerifyAsync(accountFile, loginFile.salt, loginInfo.password);
+            }).then(() => {
+                checkMfa(loginFile.mfaKey, loginInfo.currentMfa);
+
+                return accountService.deleteLoginFileAsync(hashedAccountId, loginFile.challengeId);
+            });
+        });
+    }
+
+
+    /**
+     * Need to verify the password we have on file matches
+     * what the account holder is using to log in.
+     *
+     * @param {accoutManager~accountFile} accountFile
+     * @param {string} salt
+     * @param {string} hashedPassword
+     * @return {boolean}
+     * @throws {Error} If the password hashes do not match
+     */
+    function loginVerifyAsync(accountFile, salt, hashedPassword) {
+        return hashPasswordAsync(accountFile, salt).then((passwordHash) => {
+            if (! secureHash.compare(passwordHash, base64.encode(hashedPassword))) {
+                throw new Error("Password hashes do not match");
+            }
```

Comment:
It seems to me that it might be easier to just drop these three lines in the function above.